### PR TITLE
Updates for Cairo Writer to get closer to Cairo1 syntax

### LINKED ISSUE #926: Cairo Contract Writer

### LINKED ISSUE #927: Cairo Assert Writer

### LINKED ISSUE #928: Return Writer

### LINKED ISSUE #929: Function Call Writer

### LINKED ISSUE #930: Block Writer

### LINKED ISSUE #931: Member access Writer

### LINKED ISSUE #932: Identifier Writer

### LINKED ISSUE #933: Variable declaration statement Writer

### LINKED ISSUE #934: Variable declaration Writer

### LINKED ISSUE #935: Expression statement Writer

### LINKED ISSUE #936: Parameter List Writer

### LINKED ISSUE #937: Literal Writer

### LINKED ISSUE #938: Source unit Writer

### LINKED ISSUE #939: Cairo function definition Writer

### LINKED ISSUE #940: Unchecked block Writer

### PR DESCRIPTION
**Pull request by AlejandroLabourdette**
This PR substitute #924 and target `cairo-1.0` branch

Writers updated:
- fix #939 
- fix #938
- fix #927
- fix #926
- fix #934 
- fix #931 
- fix #937 
- fix #934 
- fix #935
- fix #929
- fix #928 
- fix #930 
- fix #932 
- fix #940 
- fix #933
- fix #936

### CODE REVIEW

File: src/cairoWriter/writers/assignmentWriter.ts [hunk 1]
```diff
@@ -19,7 +19,7 @@ export class AssignmentWriter extends CairoASTNodeWriter {
       ) &&
       !(lhs instanceof TupleExpression)
     ) {
-      return [`let (${nodes[0]}) ${node.operator} ${nodes[1]};`];
+      return [`let ${nodes[0]} ${node.operator} ${nodes[1]};`];
     }
```
Comment (piwonskp) [inline src/cairoWriter/writers/assignmentWriter.ts]: It seems regardless of the condition the result is the same now, so let's remove `if`

Comment (piwonskp) [reply src/cairoWriter/writers/assignmentWriter.ts]: Also is nodes[0] same as `lhs`? Then we should use ```[`let ${nodes.lhs} ${node.operator} ${node.rhs}`]```

File: src/cairoWriter/writers/literalWriter.ts [hunk 2]
```diff
@@ -1,22 +1,11 @@
...
 
 export class LiteralWriter extends CairoASTNodeWriter {
   writeInner(node: Literal, _: ASTWriter): SrcDesc {
     switch (node.kind) {
       case LiteralKind.Number:
-        switch (primitiveTypeToCairo(node.typeString)) {
-          case 'Uint256': {
-            const [high, low] = divmod(BigInt(node.value), BigInt(Math.pow(2, 128)));
-            return [`Uint256(low=${low}, high=${high})`];
-          }
```
Comment (piwonskp) [inline src/cairoWriter/writers/literalWriter.ts:14]: Can we use literals as uint256 without explicit conversion in cairo 1?

File: src/cairoWriter/writers/expressionStatementWriter.ts [hunk 3]
```diff
@@ -15,12 +15,12 @@ export class ExpressionStatementWriter extends CairoASTNodeWriter {
...
-      node.vExpression instanceof FunctionCall &&
-      node.vExpression.kind !== FunctionCallKind.StructConstructorCall
+      (node.vExpression instanceof FunctionCall &&
+        node.vExpression.kind !== FunctionCallKind.StructConstructorCall) ||
+      node.vExpression instanceof Assignment ||
+      node.vExpression instanceof CairoAssert
     ) {
       return [[documentation, `${writer.write(node.vExpression)};`].join('\n')];
-    } else if (node.vExpression instanceof Assignment || node.vExpression instanceof CairoAssert) {
-      return [[documentation, `${writer.write(node.vExpression)}`].join('\n')];
```
Comment (piwonskp) [inline src/cairoWriter/writers/expressionStatementWriter.ts:23]: Next time please make another PR and set develop as base for cleanup / refactor changes that may be commited on the develop directly. Anything that may be merged to develop should be

Comment (piwonskp) [reply src/cairoWriter/writers/expressionStatementWriter.ts:23]: Then we can merge develop to cairo-1.0 to maintain consistency between the two

File: src/cairoWriter/writers/cairoFunctionDefinitionWriter.ts [hunk 4]
```diff
@@ -40,17 +40,16 @@ export class CairoFunctionDefinitionWriter extends CairoASTNodeWriter {
...
         .join('\n'),
     ];
   }
 
   private getDecorator(node: CairoFunctionDefinition): string[] {
-    if (node.kind === FunctionKind.Constructor) return ['@constructor'];
+    if (node.kind === FunctionKind.Constructor) return ['#[constructor]'];
     const decorators: string[] = [];
     if (node.kind === FunctionKind.Fallback) {
       decorators.push('@raw_input');
```
Comment (rjnrohit) [inline src/cairoWriter/writers/cairoFunctionDefinitionWriter.ts:55]: this decorator style also needs to be changed!

File: src/cairoWriter/writers/assignmentWriter.ts (hunk 1 above)
Comment (AlejandroLabourdette) [reply src/cairoWriter/writers/assignmentWriter.ts]: nodes[0] and nodes[1] are the string representation for lhs and rhs expressions given by the writer

File: src/cairoWriter/writers/expressionStatementWriter.ts (hunk 3 above)
Comment (AlejandroLabourdette) [reply src/cairoWriter/writers/expressionStatementWriter.ts:23]: This PR specifically transform the syntax of the output cairo file, and the changes it propose would break develop. All refractor done here was because of the structure of the new syntax (eg: changes between source unit and contract writters). Is this #962 that I create when working with bools an example of what you want?

Comment (piwonskp) [reply src/cairoWriter/writers/expressionStatementWriter.ts:23]: What I meant is the changes to `expressionStatementWriter` do not change functionality (output) so it could be done on develop. I see it now though that there is a tiny little `;` that is different between those branches.  So the output has changed and the change may actually be incompatible with cairo 0.x

File: src/cairoWriter/writers/assignmentWriter.ts (hunk 1 above)
Comment (piwonskp) [reply src/cairoWriter/writers/assignmentWriter.ts]: Oh I see, then it's good. Please take a look at the return inside and outside(below) the `if`. They seem the same, thus I guess the if is redundant

**Comment by piwonskp** [comment]
Please resolve merge conflicts

File: src/cairoWriter/writers/cairoContractWriter.ts [hunk 5]
```diff
@@ -150,3 +219,17 @@ export class CairoContractWriter extends CairoASTNodeWriter {
...
+    functionNames.push(importNode.name);
+    if (importNode.path !== imports[index + 1]?.path) {
+      // TODO: multiple imports are not avaible in cairo1
+      processedImports.push(`use ${importNode.path}::${functionNames.join(', ')}`);
+      functionNames = [];
+    }
+    return functionNames;
+  }, []);
+  return processedImports;
+}
```
Comment (piwonskp) [inline src/cairoWriter/writers/cairoContractWriter.ts]: It was deleted on #968

File: src/cairoWriter/writers/cairoContractWriter.ts [hunk 6]
```diff
@@ -71,9 +73,9 @@ export class CairoContractWriter extends CairoASTNodeWriter {
       .map((l) => (l.length > 0 ? INDENT + l : l))
       .join('\n');
 
-    const parentSourceUnit = node.parent;
-    assert(parentSourceUnit instanceof SourceUnit);
-    const otherStorageVars = parentSourceUnit.vFunctions.filter(
+    const sourceUnit = node.parent;
+    assert(sourceUnit instanceof SourceUnit, 'Contract node parent should be a Source Unit node.');
```
Comment (rodrodros) [inline src/cairoWriter/writers/cairoContractWriter.ts:77]: Please use util function `getContainingSourceUnit`

```suggestion
const sourceUnit = getContainingSourceUnit(node);
```

**Review by rodrodros — COMMENTED** [review]
It looked good to me. I think it should be merged after #968.
Also I don't see the need for #975 since it is part of this PR

File: src/cairoWriter/writers/cairoContractWriter.ts (hunk 5 above)
Comment (AlejandroLabourdette) [reply src/cairoWriter/writers/cairoContractWriter.ts]: Will be solved when imports go into cairo-1.0 branch and merge it back to this one

File: src/cairoWriter/writers/cairoContractWriter.ts (hunk 6 above)
Comment (AlejandroLabourdette) [reply src/cairoWriter/writers/cairoContractWriter.ts:77]: It will end up doing the same and this approach is a bit more strict.

Comment (rodrodros) [reply src/cairoWriter/writers/cairoContractWriter.ts:77]: More strict in what sense?

Comment (AlejandroLabourdette) [reply src/cairoWriter/writers/cairoContractWriter.ts:77]: You make sure that the contract definition is a direct child of source unit

**Review by rjnrohit — CHANGES_REQUESTED** [review]

File: src/cairoWriter/writers/cairoContractWriter.ts (hunk 6 above)
Comment (rodrodros) [reply src/cairoWriter/writers/cairoContractWriter.ts:77]: Have it your way then, I think that assertions at this stage  (the last stage of the transpiler) are just a formality to satisfy TS's type checking. I don't think a "more" strict check at this stage will do any good

**Comment by piwonskp** [comment]
Closed in favor of #983
